crates/typst/src/layout/flow/balance.rs

```diff
@@ -0,0 +1,215 @@
+use std::collections::HashMap;
+
+use crate::{
+    diag::SourceResult,
+    engine::Engine,
+    introspection::Locator,
+    layout::{Abs, Axes, Region, Regions, Size},
+};
+
+use super::{collect::Child, compose, Config as FlowConfig, Work};
+
+/// Balancing mode.
+///
+/// Dictates how the content will be balanced across available columns.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+#[repr(u8)]
+pub enum Mode {
+    /// Eagerly place items from first to last column.
+    PackStart,
+    /// Attempt equally balancing items across available columns.
+    Even,
+}
+
+trait BalancerStage<'a, 'b>: Sized {
+    type Data;
+
+    fn mode(balancer: &Balancer<'a, 'b, '_, Self>) -> Mode;
+}
+
+struct Initialized;
+impl<'a, 'b> BalancerStage<'a, 'b> for Initialized
+where
+    'a: 'b,
+{
+    type Data = Mode;
+
+    #[inline]
+    fn mode(balancer: &Balancer<'a, 'b, '_, Self>) -> Mode {
+        balancer.metadata
+    }
+}
+
+struct Measured;
+impl<'a, 'b> BalancerStage<'a, 'b> for Measured
+where
+    'a: 'b,
+{
+    type Data = ChildrenMetadata;
+
+    #[inline]
+    fn mode(balancer: &Balancer<'a, 'b, '_, Self>) -> Mode {
+        balancer.metadata.mode()
+    }
+}
```

Comment:
If you're going to be using a trait-based state machine pattern like this, I would recommend adding some comments, especially with a justfication as to why, since it is very hard to read and requires familiarity with the pattern.

However, I would probably recommend just not doing this at all?
A bool somewhere would probably suffice, but I have an intuition that you might be able to get away with deriving `default` for Mode smartly (or doing an `impl From<bool> for Mode` somehow? And maybe even `OnceCell` is applicable?

This is a lot of complication for what feels like it should be just a few functions. Although I haven't tried to rewrite it so it could be justified.

Reply:
I do see that you have a blog post on type metaprogramming (link for others: <https://tinsvagelj.net/blog/p/2024/rust_metaprogramming>), but I think it's a bit overkill for this problem.

I also went through the code a bit more closely and realized that the type states seem to only relate to this code in `flow/mod.rs`:
```rust
    let balancer = Balancer::new(&children, &config, balance);

    let mut balaced_bounds = regions.base();
    balaced_bounds.y *= Into::<usize>::into(columns) as f64;
    let mut balancer = balancer.measure(
        engine,
        locator.next(&()),
        Region::new(balaced_bounds, regions.expand),
    )?;
```
In which case, it should be possible to just push all of this into `Balancer::new` and avoid the separate states altogether?

Let me know if I'm missing something.

Reply:
I agree that it's too complicated at the moment for what it's doing, but I wanted to separate the stages of balancing without additional structs or having to clone input data.

This stage dependent data changed 2/3 times already and this just allows me to prototype faster as I'm stumbling around the existing codebase. Atm, the `ChildrenMetadata` contains all the info and `BalancerStage` can be easily turned into a struct `{Mode, Option<MeasuredData>}`.

I'll likely remove this when close to completion if I'm not consuming any expensive-to-clone inputs.